test_for_update to recognise skip_locked for MariaDB-10.6

grooverdan opened the pull request:
### Description

Alters the test for skip_locked to recognize that `SKIP LOCKED` was added to the MariaDB syntax in 10.6.0 https://mariadb.com/kb/en/select/#skip-locked.

### Checklist

This pull request is:

- [X] A minor test case fix
- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

:smile: You too :boat:

Note:

```
$ podman run  --rm  -d -e MARIADB_USER=scott -e MARIADB_PASSWORD=tiger -e MARIADB_DATABASE=test -e MARIADB_RANDOM_ROOT_PASSWORD=1  -P mariadb:10.6
7248491216e93320d7eff4c8c3a9f8c6b6c43cc84a7a65e721265616f8854f4d

$ podman port 7248491216e93320d7eff4c8c3a9f8c6b6c43cc84a7a65e721265616f8854f4d   3306
0.0.0.0:43809

$ pytest --db  mariadb  --dburi   mariadb://scott:tiger@127.0.0.1:43809/test  test/dialect/mysql/test_for_update.py
...
INTERNALERROR>   File "/home/dan/.py3/lib64/python3.9/site-packages/MySQLdb/connections.py", line 185, in __init__
INTERNALERROR>     super().__init__(*args, **kwargs2)
INTERNALERROR> sqlalchemy.exc.OperationalError: (MySQLdb._exceptions.OperationalError) (2002, "Can't connect to MySQL server on '127.0.0.1' (115)")
INTERNALERROR> (Background on this error at: https://sqlalche.me/e/14/e3q8)
```
Some queries where executed on the container instance however there's something in the test hard-coded to 3306 (observed in strace). And/or I'm doing something incorrectly.

grooverdan committed:
test_for_update to recognise skip_locked for MariaDB-10.6

Reference: https://mariadb.com/kb/en/select/#skip-locked

grooverdan commented:
hard coding port worked as expected:
```
$ pytest --db  mariadb  --dburi   mariadb://scott:tiger@127.0.0.1:3306/test  test/dialect/mysql/test_for_update.py
======================================================================================== test session starts ========================================================================================
platform linux -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/dan/.py3/bin/python3
cachedir: .pytest_cache
rootdir: /home/dan/repos/sqlalchemy, configfile: setup.cfg
plugins: cov-2.12.1
collected 26 items                                                                                                                                                                                  

test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_basic PASSED                                                                                                [  3%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_nowait PASSED                                                                                               [  7%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_basic PASSED                                                                                             [ 11%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_column_list_aliased PASSED                                                                               [ 15%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_join_aliased PASSED                                                                                      [ 19%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_join_one PASSED                                                                                          [ 23%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_nowait PASSED                                                                                            [ 26%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_read PASSED                                                                                              [ 30%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_read_nowait PASSED                                                                                       [ 34%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_read_nowait_column_list PASSED                                                                           [ 38%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_read_skip_locked PASSED                                                                                  [ 42%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_of_skip_locked PASSED                                                                                       [ 46%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_read PASSED                                                                                                 [ 50%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_read_and_nowait PASSED                                                                                      [ 53%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_read_and_skip_locked PASSED                                                                                 [ 57%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_skip_locked PASSED                                                                                          [ 61%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateCompileTest::test_for_update_textual_of PASSED                                                                                           [ 65%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_basic_lock PASSED                                                                           [ 69%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_basic_not_lock PASSED                                                                       [ 73%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_joined_lock_no_subquery PASSED                                                              [ 76%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_joined_lock_subquery PASSED                                                                 [ 80%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_joined_lock_subquery_inner_for_update PASSED                                                [ 84%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_joined_lock_subquery_inner_for_update_outer PASSED                                          [ 88%]
test/dialect/mysql/test_for_update.py::MySQLForUpdateLockingTest_mariadb+mysqldb_10_6_4_6_4::test_joined_lock_subquery_order_for_update_outer PASSED                                          [ 92%]
test/dialect/mysql/test_for_update.py::SkipLockedTest_mariadb+mysqldb_10_6_4_6_4::test_skip_locked PASSED                                                                                     [ 96%]
test/dialect/mysql/test_for_update.py::SkipLockedTest_mariadb+mysqldb_10_6_4_6_4::test_unsupported_skip_locked SKIPPED ('test/dialect/mysql/test_for_update.py::SkipLockedTest_mariadb+my...) [100%]

====================================================================================== short test summary info ======================================================================================
SKIPPED [1] test/../lib/sqlalchemy/testing/config.py:205: 'test/dialect/mysql/test_for_update.py::SkipLockedTest_mariadb+mysqldb_10_6_4_6_4::test_unsupported_skip_locked (call)' : not mysql < (8,) and not mariadb < (10, 6)
================================================================================== 25 passed, 1 skipped in 12.77s ===================================================================================
```

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision a8f9c05abea0795cfa39cd972e096e4581dbf892 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change a8f9c05abea0795cfa39cd972e096e4581dbf892: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3067

CaselIT commented:
Thanks for this pr. 

> Some queries where executed on the container instance however there's something in the test hard-coded to 3306 (observed in strace). And/or I'm doing something incorrectly.

I think the culprit is `--db  mariadb` you added that enables this db: https://github.com/sqlalchemy/sqlalchemy/blob/master/setup.cfg#L167

@zzzeek we are currently using 'mariadb(10.5.1)+pymysql' and 'mariadb(10.4.12)+mariadbconnector',

I guess we should add 10.6 (replacing 10.4?) to the test suite. If we do we should update https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/dialects/mysql/base.py#L12

grooverdan commented:
> I think the culprit is `--db mariadb`

Doh. of course - just use --dburi

> @zzzeek we are currently using 'mariadb(10.5.1)+pymysql' and 'mariadb(10.4.12)+mariadbconnector',

10.5.1 was largely a pre-release, I'm hoping no-one is really running it. My 10.5.12 tests are below:
```
$  pytest  --maxfail=100 --dburi   mariadb://bob:rhino@127.0.0.1:39109/test  --nomemory  --notimingintensive

FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_execution[<lambda>0] - ValueError: the greenlet library is required to use this function.
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_execution[<lambda>1] - ValueError: the greenlet library is required to use this function.
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_execution[<lambda>2] - ValueError: the greenlet library is required to use this function.
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_run_sync - ValueError: the greenlet library is required to use this function.
FAILED test/engine/test_pool.py::PoolEventsTest::test_checkin_event_gc[True-_exclusions0] - ValueError: the greenlet library is required to use this function.
FAILED test/engine/test_pool.py::QueuePoolTest::test_userspace_disconnectionerror_weakref_finalizer[True-_exclusions0] - ValueError: the greenlet library is required to use this function.
FAILED test/aaa_profiling/test_resultset.py::ResultSetTest_mariadb+mysqldb_10_5_12_5_12::test_fetch_by_key_legacy - AssertionError: Adjusted function call count 33594 not within 5.0% of expected ...
FAILED test/aaa_profiling/test_resultset.py::ResultSetTest_mariadb+mysqldb_10_5_12_5_12::test_fetch_by_key_mappings - AssertionError: Adjusted function call count 35598 not within 5.0% of expecte...
FAILED test/aaa_profiling/test_resultset.py::ResultSetTest_mariadb+mysqldb_10_5_12_5_12::test_raw_string - AssertionError: Adjusted function call count 26293 not within 15.0% of expected 6269, pl...
FAILED test/aaa_profiling/test_resultset.py::ResultSetTest_mariadb+mysqldb_10_5_12_5_12::test_raw_unicode - AssertionError: Adjusted function call count 26293 not within 15.0% of expected 6269, p...
FAILED test/aaa_profiling/test_resultset.py::ResultSetTest_mariadb+mysqldb_10_5_12_5_12::test_string - AssertionError: Adjusted function call count 26593 not within 15.0% of expected 6587, platfo...
FAILED test/aaa_profiling/test_resultset.py::ResultSetTest_mariadb+mysqldb_10_5_12_5_12::test_unicode - AssertionError: Adjusted function call count 26593 not within 15.0% of expected 6587, platf...
FAILED test/dialect/test_suite.py::StringTest_mariadb+mysqldb_10_5_12_5_12::test_literal_non_ascii - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 31: charac...
FAILED test/dialect/test_suite.py::TextTest_mariadb+mysqldb_10_5_12_5_12::test_literal_non_ascii - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 31: characte...
FAILED test/dialect/test_suite.py::UnicodeTextTest_mariadb+mysqldb_10_5_12_5_12::test_literal - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 50: character m...
FAILED test/dialect/test_suite.py::UnicodeTextTest_mariadb+mysqldb_10_5_12_5_12::test_literal_non_ascii - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 31: c...
FAILED test/dialect/test_suite.py::UnicodeTextTest_mariadb+mysqldb_10_5_12_5_12::test_round_trip - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 23: characte...
FAILED test/dialect/test_suite.py::UnicodeTextTest_mariadb+mysqldb_10_5_12_5_12::test_round_trip_executemany - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position ...
FAILED test/dialect/test_suite.py::UnicodeVarcharTest_mariadb+mysqldb_10_5_12_5_12::test_literal - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 50: characte...
FAILED test/dialect/test_suite.py::UnicodeVarcharTest_mariadb+mysqldb_10_5_12_5_12::test_literal_non_ascii - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 31...
FAILED test/dialect/test_suite.py::UnicodeVarcharTest_mariadb+mysqldb_10_5_12_5_12::test_round_trip - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in position 23: chara...
FAILED test/dialect/test_suite.py::UnicodeVarcharTest_mariadb+mysqldb_10_5_12_5_12::test_round_trip_executemany - UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f40d' in positi...
FAILED test/orm/test_unitofwork.py::UnicodeTest::test_basic - UnicodeEncodeError: 'charmap' codec can't encode characters in position 1-3: character maps to <undefined>
FAILED test/orm/test_unitofwork.py::UnicodeTest::test_relationship - UnicodeEncodeError: 'charmap' codec can't encode characters in position 1-3: character maps to <undefined>
================================================================ 24 failed, 13763 passed, 1898 skipped, 9 errors in 387.84s (0:06:27) ================================================================
(.py3) 
```

Using PyMySQL-1.0.2 , MariaDB-10.5.12 mariadb+pymysql://bob:rhino@127.0.0.1:39109/test results in:
```
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_execution[<lambda>0] - ValueError: the greenlet library is required to use this function.
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_execution[<lambda>1] - ValueError: the greenlet library is required to use this function.
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_execution[<lambda>2] - ValueError: the greenlet library is required to use this function.
FAILED test/ext/asyncio/test_engine_py3k.py::TextSyncDBAPI::test_sync_driver_run_sync - ValueError: the greenlet library is required to use this function.
FAILED test/engine/test_pool.py::PoolEventsTest::test_checkin_event_gc[True-_exclusions0] - ValueError: the greenlet library is required to use this function.
FAILED test/engine/test_pool.py::QueuePoolTest::test_userspace_disconnectionerror_weakref_finalizer[True-_exclusions0] - ValueError: the greenlet library is required to use this function.
```
 
> I guess we should add 10.6 (replacing 10.4?) to the test suite.

If going for 10.6, there's one test change needed as a result of 10.6 as part of the (very) slow transition (not there yet) to utf8 being mb4 by default (https://mariadb.com/kb/en/unicode/).
```
FAILED test/dialect/mysql/test_reflection.py::ReflectionTest_mariadb+mysqldb_10_6_5_6_5::test_reflection_with_table_options - AssertionError: assert 'utf8mb3' == 'utf8'
```

Side note - I had trouble seeing the 10.4 tests being run in Jenkins.

CaselIT commented:
> Side note - I had trouble seeing the 10.4 tests being run in Jenkins.

They are together in the "mysql" test. The tests can specify more than one db that will be used the te tests suite for a particular test. IIRC that run tests both mariadb and mysql 5 and 8

CaselIT commented:
Btw looking at the failed tests, most should be fixed by installing the greenlet library. The others all seem related to encoding

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3067 has been **merged**. Congratulations! :)
